microcloud/charms/src/charm.py

```diff
@@ -0,0 +1,677 @@
+            return
+        except subprocess.TimeoutExpired:
+            self.unit_blocked("This unit timed out checking its clustered status")
+            return
+
+    def _on_charm_config_changed(self, event: Union[ConfigChangedEvent, StartEvent]) -> None:
+        """React to configuration changes. (JuJu refresh)."""
+        logger.info("Updating charm config")
+
+        # Confirm that the config is valid
+        if not self.config_is_valid():
+            return
+
+        # Get all the configs that changed
+        changed = self.config_changed()
+        if not changed:
+            logger.debug("No configuration changes to apply")
+            return
+
+        # Apply all the configs that changed
+        try:
+            if (
+                "snap-channel-lxd" in changed
+                or "snap-channel-microcloud" in changed
+                or "snap-channel-microceph" in changed
+                or "snap-channel-microovn" in changed
+            ):
+                logger.info(
+                    "Changes have been detected in the snap channels, updating the snaps..."
+                )
+                self.snap_install_microcloud()
+        except RuntimeError:
+            msg = "Failed to apply some configuration change(s): %s" % ", ".join(changed)
+            self.unit_blocked(msg)
+            event.defer()
+            return
+
+    def _on_cluster_relation_joined(self, event: RelationJoinedEvent) -> None:
+        """Add a new node to the existing Microcloud cluster."""
+        if (
+            self.unit.is_leader()
+            and self.get_peer_data_str(self.unit, "clustered") == "True"
+            and event.unit
+            != self.unit  # Don't add the leader to the cluster as it is already there
+        ):
+            try:
+                self.microcloud_add()
+                logger.info("New Microcloud node successfully added")
+                return
+            except RuntimeError:
+                logger.error("Failed to add a new Microcloud node")
+                return
+
+    def _on_cluster_relation_departed(self, event: RelationDepartedEvent) -> None:
+        """Update the application information regarding the number of clustered units."""
+        if self.unit.is_leader():
+            num_clustered_units = int(self.get_peer_data_str(self.app, "num_clustered_units"))
+            self.set_peer_data_str(self.app, "num_clustered_units", str(num_clustered_units - 1))
+
+    def _on_cluster_relation_broken(self, event: RelationBrokenEvent) -> None:
+        """Effectively remove this node from the existing Microcloud cluster."""
+        if self.get_peer_data_str(self.unit, "clustered") == "True":
+            try:
+                self.microcloud_remove(os.uname().nodename)
+                logger.info("Microcloud node successfully removed")
+            except RuntimeError:
+                logger.error("Failed to remove a Microcloud node, retrying later")
```

Comment:
I think I'd do it the other way around. Instead of the removed unit trying to self remove, why not ask the leader to remove the departing unit from the cluster?

I'm suggesting that as rel-broken fires very late and you often don't have access to data bags.